architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,474 @@
+
+##### Changes to `environments` Core Resource Type
+
+A new version of `environments` will be added that does not have a hard coded list of compute kinds.
+
+```diff
++resource environment 'Applications.Core/environments@2025-05-01-preview' = {
+  name: 'myenv'
+  properties: {
+-    compute: {
+-      kind: 'kubernetes'
+-      namespace: 'default'
+-      }
+-    }
+    recipeConfig: {
+      env {
+        foo: 'bar'
+      }
+    }
+    recipes: {
+      'Applications.Datastores/redisCaches':{
+        default: {
+          templateKind: 'bicep'
+          plainHttp: true
+          templatePath: 'ghcr.io/radius-project/recipes/azure/rediscaches:latest'
+        }
+      }
+    }
+-   extensions: [
+-      {
+-        kind: 'kubernetesMetadata'
+-        labels: {
+-          'team.contact.name': 'frontend'
+-        }
+-      }
+-    ]
+  }
+}
+```
+
+> NOTE: The resource types shown here are new versions, e.g., `environments@2025-05-01-preview`, not changes to the existing versions of the resource types. This allows us to avoid breaking existing deployments.
+
+The `compute` platform is removed so that we do not have hard-coded support for specific platforms. The functionality enabled by the `compute.identity` property would be implemented via recipes so that we do not need hard-coded support for specific platforms. We could consider adding a recipe to the `environment` type if there are features enabled by `compute` that we could not achieve using recipes on the other core types.
+
+##### Changes to `applications` Core Resource Type
+
+Extensions are removed from the `applications` core type because the type of data being set in extensions would be set as recipe parameters. This is a new version of `applications`.
+
+```diff
++resource app 'Applications.Core/applications@2025-05-01-preview' = {
+  name: 'myapp'
+  properties: {
+    environment: environment
+-   extensions: [
+-     {
+-       kind: 'kubernetesNamespace'
+-       namespace: 'myapp'
+-     }
+-     {
+-       kind: 'kubernetesMetadata'
+-       labels: {
+-         'team.contact.name': 'frontend'
+-       }
+-     }
+-   ]
+  }
+}
+```
+
+#### Initialize a workspace
+
+As a platform engineer I can use `rad init` (or the equivalent rad `workspace`, `group` and `environment` commands), plus `rad recipe register`, to set up a non-Kubernetes compute platform.
+
+`rad init` would be unchanged (in terms of the user experience). Platform engineers would have to add recipes for each core type and UDT they plan to use. By default, the command will register resource types and recipes for Kubernetes provisioning.
+
+We could consider adding a flag to `rad init` that would identify a specific set of recipes. For example, `rad init aci` would install a set of default recipes for the ACI platform. However, this work is not in scope for this design.
```

Comment:
If we are going to have `rad init aci` then the current `rad init` should be `rad init kubernetes`. Or, `rad init` can ask for a platform.

Reply:
OK but can we still have `rad init` default to Kubernetes? 😺

Reply:
A separate feature spec and detailed design is planned that will cover the user experience for extensibility.